[영화 극장 선택] 심지 1, 2단계 미션 제출합니다.

sh1mj1 opened the pull request:
안녕하세요 페로로!

이번 미션 잘 부탁드립니다.

제가 미션을 시작할 때 이전 미션의 코드를 복붙 하지 않고 깃 명령어를 잘못 사용해서 제출할 때 쯤 복사, 붙여넣기를 하느라 기한에 못 맞출 뻔했습니다.
커밋 로그도 두개밖에 안 생기게 되었네요.....

정상적으로 진행되었다면 아마 [이 레포지토리](https://github.com/sh1mj1/movieTheater) 처럼 커밋이 생겼을 것입니다.
보고 싶으실 수 있으실 것 같아 남겼습니다.

https://github.com/woowacourse/android-movie-theater/assets/90040304/61d1f542-26bb-4112-ba35-7cb64189f4c1

databinding 에서 고민이 좀 많았습니다.
과연 presenter 를 xml 파일로 data binding 해주는 게 괜찮은가? 입니다.

저는 xml 은 껍데기에 가까워서 여러 뷰에서 사용될 수 있다고 생각했습니다.
마찬가지의 이유로 presenter 혹은 presenter 인터페이스를 넘겨주면 xml 과 presenter 와의 종속성이 생기게 됩니다.
그래서 presenter 를 데이터 바인딩하는 것은 데이터 객체를 바인딩하는 것보다 유연하지 않다고 생각이 들었습니다.
관련해서 페로로의 생각이 궁금합니다!

sh1mj1 committed:
feat: 이전 미션 커밋

sh1mj1 committed:
feat: 이번 미션 커밋

laco-dev commented:
> databinding 에서 고민이 좀 많았습니다.
과연 presenter 를 xml 파일로 data binding 해주는 게 괜찮은가? 입니다.
저는 xml 은 껍데기에 가까워서 여러 뷰에서 사용될 수 있다고 생각했습니다.
마찬가지의 이유로 presenter 혹은 presenter 인터페이스를 넘겨주면 xml 과 presenter 와의 종속성이 생기게 됩니다.
그래서 presenter 를 데이터 바인딩하는 것은 데이터 객체를 바인딩하는 것보다 유연하지 않다고 생각이 들었습니다.
관련해서 페로로의 생각이 궁금합니다!

엄밀히 말하면 실제 UI를 담당하는 것은 xml이 아닌 xml 내부에서 사용하는 View 관련 클래스입니다.
그렇기 떄문에 xml 상에서 프레젠터를 바인딩 하는 것 자체는 문제가 되지 않을 수 있습니다.

가령 아래 xml을 코틀린 코드로 바꾼 예시입니다.
```
// xml
<View
   android:onClick="@{() -> presenter.foo()}" />

// Kotlin
binding.view.setOnClickListener { presenter.foo() }
```

이 코드로 확인할 수 있는것은 View 인스턴스가 프레젠터에 의존하고 있지 않다는 것에요.
데이터바인딩은 컨트롤러가 View를 연결시켜주는 작업을 대신하는 과정이지 View 자체를 대신하는 역할이 아닙니다.

laco-dev commented on app/src/main/java/woowacourse/movie/domain/model/Theater.kt:
```diff
@@ -0,0 +1,11 @@
+package woowacourse.movie.domain.model
+
+data class Theater(
+    val id: Int,
+    val name: String,
+    val screens: List<Screen>,
+) {
```
하나의 극장이 N개의 상영을 가지고 있도록 구조 잘 작성 해 주셨네요. 👍🏻

laco-dev commented on app/src/main/java/woowacourse/movie/domain/model/Theater.kt:
```diff
@@ -0,0 +1,11 @@
+    val name: String,
+    val screens: List<Screen>,
+) {
+    fun isScreening(screen: Screen): Boolean = screens.contains(screen)
```
무언가를 포함하고 있다는 기능은 isXXX() 보다는 hasXXX()를 주로 사용하고는 합니다.

sh1mj1 replied:
`상영하고 있다` 라는 의미로 사용했습니다.
하지만 hasScreen 이 더 나아보이네요!

laco-dev commented on app/src/main/java/woowacourse/movie/domain/model/Theater.kt:
```diff
@@ -0,0 +1,11 @@
+) {
+    fun isScreening(screen: Screen): Boolean = screens.contains(screen)
+
+    fun allScreeningTimeCount(screen: Screen): Int = screen.allScreeningTime(WeeklyScreenTimePolicy()).count()
```
극장이 주말시간에만 의존하고 있는 구조가 정상일까요?

sh1mj1 replied:
이 부분의 설계가 이상하다고 생각했었습니다.
더 고민해서 고쳐보겠습니다!

laco-dev commented on app/src/main/java/woowacourse/movie/domain/model/Theaters.kt:
```diff
@@ -0,0 +1,12 @@
+data class Theaters(val theaters: List<Theater>) {
+    constructor(vararg theaters: Theater) : this(theaters.toList())
+
+    fun screeningTheater(screen: Screen): Theaters =
+        Theaters(
+            theaters.filter { theater ->
+                theater.screens.any { it.id == screen.id }
+            },
+        )
```
상영이 여러개의 극장에 공존할 수 있는 개념일까요?
영화를 기준으로 찾는다면 이해가 되는데, 상영을 통해 극장들을 역산하는 경우가 있을까 생각이 들어서요

sh1mj1 replied:
같은 상영이 서로 다른 극장에 있을 수 없군요..
여러 영화관에서 똑같은 상영을 할 수 있다고 생각했는데, 제가 잘못생각했네요.
애초에 모든 좌석을 찾을 때, 상영을 기준으로 좌석을 찾는데, 서로 다른 극장에서 같은 영화를 상영하지만 좌석이 다른 게 일반적이니까요.

laco-dev commented on app/src/main/java/woowacourse/movie/ui/Currency.kt:
```diff
@@ -7,18 +7,18 @@ import java.util.Locale
+        private val koreaCurrencyFormat = DecimalFormat("#,###원")
 
-        private var MU_FORMAT: MutableMap<String, NumberFormat> =
+        private var muFormat: MutableMap<String, NumberFormat> =
```
muFormat은 무슨 뜻인가요?
정말 궁금해서 물어봅니다.

sh1mj1 replied:
엇,  ⇧F6 으로 이름을 변경하다가 잘 못 변경된 것 같네요
`currencyFormats` 환율 형식이 맞습니다

laco-dev commented on app/src/main/java/woowacourse/movie/ui/Currency.kt:
```diff
@@ -7,18 +7,18 @@ import java.util.Locale
-            if (MU_FORMAT[country] == null) {
-                MU_FORMAT[country] = NumberFormat.getCurrencyInstance(Locale.getDefault())
+            if (muFormat[country] == null) {
+                muFormat[country] = NumberFormat.getCurrencyInstance(Locale.getDefault())
```
country가 제가 생각한 나라 관련 내용이 맞나요?
그런데 실제로는 키값으로 구분하기 위한 개념으로만 사용하지 Locale.getDefault()로만 만들어지지 않나요?

sh1mj1 replied:
네, 맞습니다
금액에 대해서 어디선가 나라 관련 내용을 준다면, 금액이 어떻게 바뀔지 이전 movie-ticket 미션 step1 에서 생각하다가 구성했던 것 같습니다.
실제로는 주요하게 동작하지는 않고 있어요..

laco-dev commented on app/src/main/java/woowacourse/movie/ui/MainActivity.kt:
```diff
@@ -0,0 +1,53 @@
+    override fun onCreate(savedInstanceState: Bundle?) {
+        super.onCreate(savedInstanceState)
+
+        supportFragmentManager.commit {
+            setReorderingAllowed(true)
+            add(R.id.fragment_container_view, ReservationHistoryFragment())
+        }
```
여러 탭을 이동하면서, 화면을 회전해보세요.
그리고 어떤 결과가 나오는지 또한 supportFragmentManager.fragments의 로그를 출력해봤을 떄 어떻게 나오는지도 알려주세요

sh1mj1 replied:
생명 주기 함수들에서 supportFragmentManager.fragments 를 출력해보았습니다.

![image](https://github.com/woowacourse/android-movie-theater/assets/90040304/02a41ec8-c9b6-4a56-96cd-ba66936c802c)
이렇게 FragmentManager 에 fragment 가 계속 추가되고 있네요.

현재 MainActivity 에서 onCreate 될 때 항상 FragmentManager 에 ReservationHistoryFragment() 를 추가해 주고 있는 것이 문제인 것 같습니다.

간단히 activity 에서 savedInstanceState 의 null 여부를 체크해주면 될 것 같아요.

laco-dev commented on app/src/main/java/woowacourse/movie/ui/MainActivity.kt:
```diff
@@ -0,0 +1,53 @@
+                }
+
+                R.id.setting -> {
+                    supportFragmentManager.commit {
+                        setReorderingAllowed(true)
+                        replace(R.id.fragment_container_view, SettingFragment())
+                    }
```
탭을 이동할 때 마다 매번 새로운 프래그먼트 인스턴스를 만들어서 넣으면 이전에 봤던 화면은 어떻게 되는 건가요?

sh1mj1 replied:
새 fragment를 계속 생성하면 GC 의 부담이 커지겠네요.
또 새 인스턴스가 생성되다 보니 사용자가 이전에 보았던 화면 스크롤 위치 같은 것들이 없어질 것 같아요.

1. 액티비티가 fragment 들을 인스턴스 변수로 가지고 있는 것도 고려해볼 수 있을 것 같네요!
2. 아니면... FragmentManager.fragments 에 미리 추가해두고, 화면에 따라 프래그먼트들이 보이고, 감춰지도록 할 수도 있을 것 같아요

일단 1 번 방법으로 해보겠습니다

laco-dev commented on app/src/main/java/woowacourse/movie/ui/UiMapping.kt:
```diff
@@ -34,3 +39,51 @@ fun Screen.toDetailUI(image: Image<Any>) =
+    textView: TextView,
+    price: Int,
+) {
+    textView.text = String.format("%s(현장 결제)", Currency.of(Locale.getDefault().country).format(price))
```
왜 이 출력은 strings  리소스를 활용하지 않았을까요?

sh1mj1 replied:
strings 리소스에 접근하려면 textview 의 context 에 접근해야 해서, `textView.context` 와 같이 접근하는 것이 괜찮은가에 대해 고민하다 이렇게 둔 것 같습니다.
이 부분 더 고민해볼게요!

laco-dev replied:
모든 View는 Conext를 가지고 있어요. TextView도 View입니다.

laco-dev commented on app/src/main/java/woowacourse/movie/ui/UiMapping.kt:
```diff
@@ -34,3 +39,51 @@ fun Screen.toDetailUI(image: Image<Any>) =
 
 fun DateRange.toUi(): String = "$start ~ $endInclusive"
+
+@BindingAdapter("app:textFormat")
+fun textUiFormat(
+    textView: TextView,
+    seats: List<Seat>,
+) {
+    textView.text =
+        seats.joinToString(
+            separator = ",",
+            postfix = " |",
+        ) { "${'A' + it.position.row}${it.position.col + 1}" }
+}
```
최상위에 작성된 데이터바인딩은 모든 화면 모든 코드에서 접근이 가능하기 때문에 매우 조심해야 합니다.

만약 다른 화면에서도 똑같이 textFormat으로 바인딩 어댑터를 만들면 어떤 문제가 생길까요?
그렇다면 이러한 문제를 해결하려면 어떤 전략을 사용할 수 있을까요?

sh1mj1 replied:
@BindingAdapter 를 똑같은 이름으로 다른 화면에서 재정의하면, 네임스페이스 충돌이 일어날 수 있겠네요.
또 관리하기도 좀 어려울 것 같아요.

seats:textFormat 이런 식으로 prefix 를 설정해줄 수 있을 것 같아요.

laco-dev replied:
prefix?? 네임스페이스라고 한다면 그것은 구분하기 위한 방법은 아니에요.
피드백 강의에 이 내용이 포함되있을 것 같은데 네임스페이스는 아무 값이나 해도 돌아갑니다.

laco-dev commented on app/src/main/java/woowacourse/movie/ui/UiMapping.kt:
```diff
@@ -34,3 +39,51 @@ fun Screen.toDetailUI(image: Image<Any>) =
+    textView.text = dateRange.toUi()
+}
+
+@BindingAdapter("app:srcScreenPoster")
+fun screenPoster(
+    imageView: ImageView,
+    image: Image<Any>,
+) {
+    imageView.setImageResource(image.imageSource as Int)
+}
```
바인딩 어댑터를 UiMapping 이라는 파일에 모두 모아주셨네요.

만약 화면이 200개, 300개가 되면 이 파일은 어떻게 될까요?

sh1mj1 replied:
너무 많은 메서드가 여기에 집중되겠네요.

여러 바인딩 어댑터로 나눠 보겠습니다.

laco-dev commented on app/src/main/java/woowacourse/movie/ui/detail/ScreenDetailActivity.kt:
```diff
@@ -69,7 +72,7 @@ class ScreenDetailActivity : AppCompatActivity(), ScreenDetailContract.View {
 
                 override fun reserve(screenId: Int) {
-                    presenter.reserve(screenId)
+                    presenter.reserve(screenId, theaterId)
```
상세 화면이 생성되면서 프레젠터가 아이디 값을 같이 가질 수 있지 않을까요?
한번 생성 이후에는 더이상 변경되지 않을 값 처럼 보이는데, 여러 로직에서 매번 아이디를 주고 받는 점이 불필요해 보입니다.

sh1mj1 replied:
프레젠터가 인스턴스 변수로 id 값을 저장하도록 해보겠습니다!

laco-dev commented on app/src/main/java/woowacourse/movie/ui/detail/ScreenDetailActivity.kt:
```diff
@@ -112,15 +115,18 @@ class ScreenDetailActivity : AppCompatActivity(), ScreenDetailContract.View {
     override fun showDateTimePicker(
         dateRange: DateRange,
-        screenTimePolicy: ScreenTimePolicy,
+        screenTimePolicy: WeeklyScreenTimePolicy,
```
요기도 주말이랑 의존이 되어버렸네요

laco-dev commented on app/src/main/java/woowacourse/movie/ui/detail/view/ScreenDetailDateTimeSpinnerView.kt:
```diff
@@ -2,161 +2,84 @@ package woowacourse.movie.ui.detail.view
+        DataBindingUtil.inflate(LayoutInflater.from(context), R.layout.holder_spinner_date_time, this, true)
     }
 
+    private lateinit var dateAdapter: DateAdapter
+    private lateinit var timeAdapter: TimeAdapter
```
show가 호출되면 매번 새로운 어댑터가 만들어지는데 프로퍼티로 관리될 필요가 있나요?

sh1mj1 replied:
아, 로직 상, 이 화면이 보여질 때 show 가 호출되어서 어댑터가 한 번 만들어지네요.
말씀대로 프로퍼티로 관리할 필요가 없을 것 같습니다!

laco-dev commented on app/src/main/java/woowacourse/movie/ui/detail/view/ScreenDetailScreenView.kt:
```diff
@@ -2,32 +2,19 @@ package woowacourse.movie.ui.detail.view
-    private val description: TextView by lazy { findViewById(R.id.tv_description) }
-    private val poster: ImageView by lazy { findViewById(R.id.iv_poster) }
-    private val dateRange: TextView by lazy { findViewById(R.id.tv_screen_date) }
+    private val binding: HolderScreenDetailScreenBinding =
+        DataBindingUtil.inflate(LayoutInflater.from(context), R.layout.holder_screen_detail_screen, this, true)
```
ScreenView라는 이름이 무엇을 위한 뷰인지 잘 예측이 안되어서요.

이름만으로 본다면 화면 전체를 그리는 View인건가요?
그건 액티비티나 프래그먼트처럼 보이는데요

sh1mj1 replied:
![image](https://github.com/woowacourse/android-movie-theater/assets/90040304/930b1885-8dc2-4cfb-ba73-c44315941d05)

위처럼 상영 상셰 화면에서 상영작의 구체적인 정보를 표시하는 부분입니다.
네이밍을 어떻게 해야 예측하기 쉬울까요.... 

인터페이스의 이름은 `ScreenDetailInformationView`, 그리고 구현체의 이름은 `DefaultScreenDetailInformationView` 로 바꾸었습니다

laco-dev commented on app/src/main/java/woowacourse/movie/ui/detail/view/adapter/DateAdapter.kt:
```diff
@@ -0,0 +1,41 @@
+class DateAdapter(
+    context: Context,
+    dateRange: DateRange,
+    val updateTime: (LocalDate) -> Unit,
+    val selectDate: (Int) -> Unit,
```
이 두개의 고차함수가 어떤 역할을 하는지 잘 이해가 안되어서요.

1. updateTime(LocalDate)가 의미하는 것이 오히려 선택이 아닌가요?
2. selectDate()가 왜 Int를 넘겨주는 것인지, 이 값은 그러면 무엇인지 모르겠어요

sh1mj1 replied:
updateTime 에서는 선택한 날짜로 TimeAdapter 의 시간이 수정되는 것이고
selectDate 는 선택한 날짜의 위치 정보를 가지고 presenter 로 넘겨주는 것입니다.

이 부분 구현 수정 하겠습니다

laco-dev commented on app/src/main/java/woowacourse/movie/ui/detail/view/ScreenDetailTicketView.kt:
```diff
@@ -2,46 +2,41 @@ package woowacourse.movie.ui.detail.view
-        plusBtn.setOnClickListener {
+        binding.btnPlus.setOnClickListener {
             ticketReserveListener.increaseTicket()
         }
-        minusBtn.setOnClickListener {
+        binding.btnMinus.setOnClickListener {
             ticketReserveListener.decreaseTicket()
         }
-        reserveBtn.setOnClickListener {
+        binding.btnReserveDone.setOnClickListener {
```
전체 코드 변경을 확인하지 않아서 확인차 질문 드립니다.
TicketView의 인터페이스가 실제로 인터페이스로 활용이 되고 있나요?

ScreenDetailTicketView에 똑같이 public 함수로 만들어도 동일한 것 처럼 보여서요

sh1mj1 replied:
이 뷰를 사용하는 액티비티에서 아래처럼 타입을 인터페이스로 두려고 했는데 구현체로 되어 있었네요.
```kotlin
private val screenDetailView: ScreenDetailInformationView by lazy { findViewById(R.id.screen_detail_screen_view) }
private val ticketView: TicketView by lazy { findViewById(R.id.screen_detail_ticket_view) }
private val dateTimeSpinnerView: DateTimeSpinnerView by lazy { findViewById(R.id.screen_detail_date_time_spinner_view) }
```

이 부분 수정하겠습니다.

laco-dev commented on app/src/main/java/woowacourse/movie/ui/detail/view/ScreenDetailDateTimeSpinnerView.kt:
```diff
@@ -2,161 +2,84 @@ package woowacourse.movie.ui.detail.view
+        )
+        initTimeAdapter(dateRange.start, screenTimePolicy, selectTimeListener)
+
+        binding.spnDate.onItemSelectedListener = dateAdapter.initClickListener()
+        binding.spnTime.onItemSelectedListener = timeAdapter.initClickListener()
```
initClickListener를 어디서는 초기화하는데 쓰이고 어디서는 반환하는데 쓰이고 일관성이 떨어지지 않을까요?

만능 함수처럼 그냥 클릭리스너면 전부 initClickListener라고 지어야지? 같은 느낌

sh1mj1 replied:
`initClickListener` 라는 네이밍의 메서드를 대부분의 사용처에서는 Unit 인 메서드로 사용해 놓고, 여기서는 리턴 타입이 있는 것으로 사용하고 있어 통일성이 없네요...

사실 `dateAdapter` 가 `AdapterView.OnItemSelectedListener` 를 구현하도록 만드려고 했습니다.
이를 구현하는 도중 잠시, 네이밍을 이렇게 해 두었다가,  무언가 잘 안되어서 넘어갔었는데,
네이밍을 변경하는 것을 빼먹었습니다.

이것도 좋지 않은 습관인 것 같네요..

laco-dev commented on app/src/main/java/woowacourse/movie/ui/home/HomeFragment.kt:
```diff
@@ -0,0 +1,69 @@
+        )
+    }
+
+    override fun onCreateView(
+        inflater: LayoutInflater,
+        container: ViewGroup?,
+        savedInstanceState: Bundle?,
+    ): View {
```
onCreateView()와 onViewCreated()는 어떤 차이가 있나요?

다른 일반적인 프래그먼트 로직을 작성할 떄 주로 각 함수는 어떤 역할을 할까요?

sh1mj1 replied:
공부해보니 아래와 같았습니다.
* `onCreateView`
  * 레이아웃 파일을 inflateing 해서 뷰 계층 구성. 프래그먼트의 뷰 계층을 리턴.
  * view 계층을 생성하는 데만 집중하고 뷰 초기화는 이후 단계에서 수행하는 것이 좋음

* `onViewCreated`
  * 저장된 상태가 있다면 뷰에 복구.
  * 프래그먼트의 뷰가 생성된 이후의 초기화 작업을 수행.

그에 맞게 수정했습니다!

laco-dev commented on app/src/main/java/woowacourse/movie/ui/home/HomeFragment.kt:
```diff
@@ -0,0 +1,69 @@
+                { screenId -> homePresenter.loadTheaters(screenId) },
+                { adId -> Toast.makeText(activity, "광고 클릭 id: $adId", Toast.LENGTH_SHORT).show() },
+            )
+        binding.adapter = adapter
```
바인딩으로 어댑터를... 직접 넣는경우는 잘 없어요. 
일반적으로는 리사이클러뷰에 접근해서 어댑터를 설정 해줍니다.

sh1mj1 replied:
그렇군요.. 

어댑터에 동적으로 설정하는 기능이 많기도 하고,
리사이클러뷰에 접근해서 어댑터를 프래그먼트나 액티비티에서 하는 코드가 딱히 길지 않아서 그런 건가요?

laco-dev replied:
아뇨, 어댑터를 할당하는 것이 데이터바인딩의 목적이 아니니까요.
그런 맥락이라면 뷰의 모든 기능을 전부 데이터바인딩으로 돌릴 수 있습니다.

데이터를 -> 바인딩 한다 라는 관점이기 때문에 어댑터는 따로 데이터로 보지는 않아요.
이 내용은 MVVM 라이브데이터를 학습하시고 다시 살펴보시면 좋을 것 같아요

laco-dev commented on app/src/main/java/woowacourse/movie/ui/home/HomeFragment.kt:
```diff
@@ -0,0 +1,69 @@
+        screen: Screen,
+        theaters: Theaters,
+    ) {
+        val theaterAdapter =
+            TheaterAdapter(screen) { screenId, theaterId ->
+                ScreenDetailActivity.startActivity(requireContext(), screenId, theaterId)
+            }
+
```
극장 어댑터는 왜 매번 새로운 어댑터를 만들어서 처리하도록 하였나요?

sh1mj1 replied:
극장 어댑터의 아이템(극장)을 선택했을 때 `ScreenDetailActivity` 가 실행되어야 하는데, 이 때 상영 정보와 영화관 정보를 넘겨주어야 합니다. 그래서 TheaterAdapter 가 생성자로 Screen 을 가지고 있어서 항상 새 어댑터를 만들게 되었습니다.
새 어댑터를 만들지 않는 방식으로 변경해보겠습니다.

laco-dev commented on app/src/main/java/woowacourse/movie/ui/home/HomeFragment.kt:
```diff
@@ -0,0 +1,69 @@
+        val theaterBottomSheet = TheaterBottomSheet(theaterAdapter)
+
+        theaterAdapter.submitList(theaters.theaters)
+        theaterBottomSheet.show(parentFragmentManager, "theaterBottomSheet")
```
parentFragmentManager와 childFragmentManager 두개가 있어요. 각각 어떤 차이가 있을까요?

sh1mj1 replied:
* `parentFragmentManager`
  * 프래그먼트의 상위 계층을 관리하는, 프래그먼트를 담고 있는 액티비티나 상위 프래그먼트 매니저.
* `childFragmentManager`
  * 프래그먼트가 다른 프래그먼트를 내부에 중첩해서 사용해야 할 경우 관리하는 프래그먼트 매니저.
  * 특정 프래그먼트 내에서 독립적인 프래그먼트를 관리함.

TheaterBottomSheet 는 HomeFragment 내에서만 사용되기 때문에 childFragmentManager 를 사용하는 게 맞겠네요!
![image](https://github.com/woowacourse/android-movie-theater/assets/90040304/c638128b-b68f-49fa-b456-7a81d10693e5)

laco-dev commented on app/src/main/java/woowacourse/movie/ui/home/HomePresenter.kt:
```diff
@@ -0,0 +1,44 @@
+    private val theaterRepository: TheaterRepository = DummyTheaters(),
+) : HomeContract.Presenter {
+    override fun loadScreen() {
+        val screens =
+            screenRepository.load()
+                .map { screen -> screen.toPreviewUI(image = movieRepository.imageSrc(screen.movie.id)) }.toMutableList()
```
가독성이 다소 떨어집니다.
별도 변수를 분리하는 방법 등으로 조금 개선 해주세요

laco-dev commented on app/src/main/java/woowacourse/movie/ui/home/HomePresenter.kt:
```diff
@@ -0,0 +1,44 @@
+
+        val list = mutableListOf<ScreenAd>()
+
+        (1..screens.size + screens.size / 3).map {
+            if (it % 4 == 0) {
+                list.add(ad)
+            } else {
+                list.add(screens.removeFirst())
+            }
+        }
```
map 대신에 mapIndexed를 활용하면 리스트를 제거할 필요 없이 만들면서 아이템을 추가할 수 있을 것 같네요

sh1mj1 replied:
그렇군요 감사합니다!

laco-dev commented on app/src/main/java/woowacourse/movie/ui/home/TheaterBottomSheet.kt:
```diff
@@ -0,0 +1,28 @@
+import woowacourse.movie.R
+import woowacourse.movie.ui.home.adapter.TheaterAdapter
+
+class TheaterBottomSheet(private val theaterAdapter: TheaterAdapter) : BottomSheetDialogFragment() {
```
바텀시트 프래그먼트도 프래그먼트이기 때문에, 생성자로 값을 넣어주면 안됩니다.
생성자에 값을 넣으면 왜 안되는지는 찾아보세요!!

sh1mj1 replied:
프래그먼트를 재생성할 때 시스템이 관리하는 Bundle 객체를 통해 복원되는데, 생성자에 전달된 값은 복원되지 않는 군요.
그러면 프래그먼트를 재생성할 때 값이 사라질 것 같아요.

HomeFragment 에서 TheaterBottomSheet 를 보여주는 책임만 갖고, 
TheaterBottomSheet 에서 ScreenDetailActivity 로 이동하도록 리팩터링하면 여러 문제가 해결될 것 같네요.
매번 극장 어댑터를 새로 만드는 문제도 포함해서요!

laco-dev commented on app/src/main/java/woowacourse/movie/ui/home/TheaterBottomSheet.kt:
```diff
@@ -0,0 +1,28 @@
+        savedInstanceState: Bundle?,
+    ) {
+        super.onViewCreated(view, savedInstanceState)
+        view.findViewById<RecyclerView>(R.id.rv_theater).adapter = theaterAdapter
```
이 화면도 데이터바인딩을 동일하게 사용 해주세요

laco-dev commented on app/src/main/java/woowacourse/movie/ui/home/adapter/AdvertisementViewHolder.kt:
```diff
@@ -0,0 +1,18 @@
+import woowacourse.movie.domain.model.ScreenAd
+
+class AdvertisementViewHolder(
+    private val onItemClick: (id: Int) -> Unit,
```
아이템클릭인데 전달하는 파라미터가 Int 타입으로 줄 필요가 있나요?

sh1mj1 replied:
클릭된 advertisement 에 따라 후의 행동이 달라진다면 advertisement 의 id 가 필요할 것 같아서 이렇게 구현했습니다.

laco-dev replied:
Advertisement 타입을 그대로 전달하는 것은 안되나요?

laco-dev commented on app/src/main/java/woowacourse/movie/ui/home/adapter/ScreenAdapter.kt:
```diff
@@ -0,0 +1,63 @@
+    private val onScreenClick: (id: Int) -> Unit,
+    private val onAdClick: (id: Int) -> Unit,
+) : ListAdapter<ScreenAd, RecyclerView.ViewHolder>(ScreenPreviewUiDiffUtil()) {
+    private lateinit var inflater: LayoutInflater
```
Inflater를 캐시하면 어떤 장점이 있나요?

실제로 무거운 작업은 inflater를 가져오는게 아니고 실제 inflate 하는 작업입니다.

sh1mj1 replied:
그렇군요!
그렇다면 굳이 인스턴스 변수를 하나 더 가질 필요가 없겠네요..

laco-dev commented on app/src/main/java/woowacourse/movie/ui/home/adapter/ScreenAdapter.kt:
```diff
@@ -0,0 +1,63 @@
+) : ListAdapter<ScreenAd, RecyclerView.ViewHolder>(ScreenPreviewUiDiffUtil()) {
+    private lateinit var inflater: LayoutInflater
+
+    override fun getItemViewType(position: Int): Int =
+        when {
+            ((position + 1) % ADVERTISEMENT_INTERVAL == 0) -> ScreenType.ADVERTISEMENT.id
+            else -> ScreenType.SCREEN.id
+        }
```
ListAdapter를 사용하면서, 아이템을 통해 타입을 분리한다면 sealed class를 활용해 어댑터를 만드는 방법을 한번 찾아보셔도 좋습니다.

꼭 반영하지 않아도 됩니다.

sh1mj1 replied:
말씀하신 대로 PR 올리기 전에 페어와 만들다가 실패했는데 다시 공부해서 적용했습니다!
```kotlin
sealed class ScreenAdViewHolder(view: View) : RecyclerView.ViewHolder(view) {
    abstract fun bind(screenAd: ScreenAd)

    class ScreenViewHolder(
       // ..
    ) : ScreenAdViewHolder(binding.root) {
        override fun bind(screenAd: ScreenAd) {
           // ...
        }
    }

    class AdvertisementViewHolder(
       // ..
    ) : ScreenAdViewHolder(binding.root) {
        override fun bind(screenAd: ScreenAd) {
            // ...
        }
    }
}
```

```kotlin
class ScreenAdapter(
    private val onScreenClick: (id: Int) -> Unit,
    private val onAdClick: (id: Int) -> Unit,
) : ListAdapter<ScreenAd, ScreenAdViewHolder>(ScreenPreviewUiDiffUtil()) {
    // ...
}
```
이런 식으로 적용하면 되겠네요!

이렇게 구현하려다가 실패한 흔적들(ScreenAd 라는 sealed interface)도 개선했습니다.

laco-dev replied:
좋습니다. 다만 리스트어댑터가 교육과정상 충분히 리사이클러뷰 어댑터를 학습하고 난뒤의 과정이라고 하네요!
아마 다음 미션에서는 리사이클러뷰 어댑터를 잘 쓰도록 해야 할 수도 있어요 😢

laco-dev commented on app/src/main/java/woowacourse/movie/ui/home/adapter/ScreenAdapter.kt:
```diff
@@ -0,0 +1,63 @@
+    ) {
+        val item = getItem(position)
+
+        if (item is ScreenAd.ScreenPreviewUi) {
+            (holder as ScreenViewHolder).bind(item)
+        } else if (item is ScreenAd.Advertisement) {
+            (holder as AdvertisementViewHolder).bind(item)
+        }
```
코틀린에서는 else if 보다 선호되는 when을 사용해주세요 :)

sh1mj1 replied:
넵 그게 좋겠네요!!

laco-dev commented on app/src/main/java/woowacourse/movie/ui/home/adapter/ScreenAdapter.kt:
```diff
@@ -0,0 +1,63 @@
+    }
+
+    companion object {
+        private const val ADVERTISEMENT_INTERVAL = 4
```
실제 광고가 껴들어가는 시점은 프레젠터에서 결정하는데 어댑터의 이 값이 유의미할까요?

laco-dev commented on app/src/main/java/woowacourse/movie/ui/seat/SeatReservationContract.kt:
```diff
@@ -6,31 +6,48 @@ import woowacourse.movie.domain.model.TimeReservation
 interface SeatReservationContract {
     interface View {
-        fun showSeats(seats: Seats)
+        fun initBinding(
+            totalPrice: Int,
+            timeReservation: TimeReservation,
+        )
```
프레젠터와 마찬가지로 뷰가 구체적으로 어떤 동작을 할지는 뷰가 결정하도록 해야 합니다.
즉 initBiniding()처럼 뷰가 바인딩을 초기화 할지 뷰를 초기화 할지 화면을 띄울지 등은 View가 할 역할이라는 의미가 됩니다.

show() 와 같이 기능을 위주로 인터페이스를 만들어주세요

sh1mj1 replied:
그렇군요 
인터페이스의 메서드 이름이 더 추상화되고 선언적으로 변경해보겠습니다

laco-dev commented on app/src/main/java/woowacourse/movie/ui/seat/SeatReservationContract.kt:
```diff
@@ -6,31 +6,48 @@ import woowacourse.movie.domain.model.TimeReservation
 
         fun showSeatReservationFail(throwable: Throwable)
 
+        fun showDialog(
+            reservationId: Int,
+            theaterId: Int,
+        )
```
마찬가지로 다이얼로그가 뜰지 아닐지는 뷰만이 알고 있습니다.

laco-dev commented on app/src/main/java/woowacourse/movie/ui/seat/SeatReservationContract.kt:
```diff
@@ -6,31 +6,48 @@ import woowacourse.movie.domain.model.TimeReservation
         )
 
-        fun reserve(screenId: Int)
+        fun attemptReservation(
+            screenId: Int,
+            theaterId: Int,
+        )
+
+        fun completeReservation(
+            screenId: Int,
+            theaterId: Int,
+        )
```
동적으로 변경되거나 생성되는 값이 아닌데 이 아이디 값은 프레젠터가 생성되면서 동시에 가지고 있도록 할 수 있지 않을까요?

sh1mj1 replied:
넵 다른 Contract 와 더불어서 수정하겠습니다

laco-dev commented on app/src/main/java/woowacourse/movie/ui/seat/SeatReservationPresenter.kt:
```diff
@@ -1,60 +1,50 @@
-        if (selectedSeats.contains(seat)) {
-            seatView.isSelected = !seatView.isSelected // 선택 상태 토글
-            selectedSeats.remove(seat)
+        val seat = loadedAllSeats.findSeat(position)
```
selectSeat() 함수가 다소 복잡합니다.
함수 분리 등의 방법을 이용해 가독성을 개선 해주세요

laco-dev commented on app/src/main/res/layout-land/activity_seat_reservation.xml:
```diff
@@ -0,0 +1,129 @@
+                <TextView
+                    android:id="@+id/tv_seat_reservation_screen"
+                    android:layout_width="400dp"
+                    android:layout_height="30dp"
```
텍스트뷰 같은 경우는 화면이 작아지거나 글자가 많아지거나 등 텍스트가 길어질 수 있는데요.
이 때 높이를 30dp와 같이 고정하게 된다면, 치명적인 문제가 발생하고는 합니다.

이 텍스트 뷰의 텍스트에 5줄 이상의 텍스트를 넣으면 어떻게 되는지 알려주세요

sh1mj1 replied:
엇 높이가 30dp 로 고정되어 있군요..

![image](https://github.com/woowacourse/android-movie-theater/assets/90040304/d02ef11e-b49f-4ad9-a832-8c48d610c096)
이런 식으로 텍스트가 잘리게 되네요..

wrap_content 로 설정해서 크기를 자동으로 조정하게 만들어야 겠네요!

laco-dev commented on app/src/main/res/layout-land/activity_seat_reservation.xml:
```diff
@@ -0,0 +1,129 @@
+                tools:text="20,000원" />
+        </androidx.constraintlayout.widget.ConstraintLayout>
+
+        <androidx.appcompat.widget.AppCompatButton
```
`<androidx.appcompat.widget.AppCompatButton />` 과 `<Button />`는 어떤 차이가 있나요?

sh1mj1 replied:
여기서 Button 을 쓰는 것과 AppCompatButton 을 쓰는 것이 차이가 없네요
최상위 액티비티나 다이얼로그가 appcompat 으로 제공되면, 즉, AppCompatActivity 등을 상속받으면, 자동으로 AppCompateButton 을 사용하게 되네요!

뿐만 아니라 TextView, ImageView, ImageButton, EditText 등 모두 그렇다는 것을 처음 알았네요.

AppCompatButton 은 Button 클래스를 상속해서 추가적인 기능과 호환성을 제공한다고만 알고 있었는데 감사합니다!

laco-dev commented on app/src/main/res/layout-land/holder_screen_detail_screen.xml:
```diff
@@ -1,104 +1,120 @@
+                    android:id="@+id/tv_screen_running_time"
+                    android:layout_width="wrap_content"
+                    android:layout_height="wrap_content"
+                    android:text="@{String.format(@string/running_time, screenDetailUi.movieDetailUI.runningTime)}"
```
string resource에 대해서도 데이터바인딩을 지원합니다.
```
"@{@string/running_time(...)}"
```

sh1mj1 replied:
수정하겠습니다. 
xml 이 조금더 깔끔해지겠네요!

laco-dev commented on app/src/test/java/woowacourse/movie/ui/screen/HomePresenterTest.kt:
```diff
@@ -0,0 +1,47 @@
+
+    @BeforeEach
+    fun setUp() {
+        view = mockk<HomeContract.View>()
```
relaxed = true 로 생성하면 불필요한 just runs 같은 동작을 기본 값으로 처리할 수 있습니다.

laco-dev commented on app/src/test/java/woowacourse/movie/ui/screen/HomePresenterTest.kt:
```diff
@@ -0,0 +1,47 @@
+
+        presenter.loadTheaters(1)
+
+        verify { view.showTheaters(any(), any()) }
```
엄격히 간다면, 어떤 값이 호출되었는지 any()로 확인하지 않고 실제 Mock 데이터를 비교하는 것이 일반적입니다.
다만 이 과정에서는 그렇게 까지 작성하지 않아도 되어서 참고용으로 남깁니다.

sh1mj1 replied:
넵 사실은 레포지토리 같은 경우 fake 객체를 만들어서 프레젠터 생성자에 넣어  테스트를 하다가 이번 페어 프로그래밍 때는 이렇게 하게 되었네요.

나중에 변경해보겠습니다.

laco-dev reviewed (changes requested):
커스텀 뷰를 이용한 구성이 인상깊었습니다.
다만 아직은 커스텀 뷰에 대한 자세한 원리를 다루지 않았기 때문에 단순히 코드만 별도 클래스로 옮긴 정도의 느낌은 들었어요.

나중에 실제로 다루게 될 시점에 자연스럽게 알아가실 것 같아요.
작업량이 기본적으로 많아서 모든 부분을 확인하지는 못했습니다.

이미 안드로이드 개발 경험이 충분히 있으신 것 같아서, 구조적인 부분 위주로 남겼습니다.

sh1mj1 committed:
refactor(Theater): 메서드 이름 변경

- Before: isScreening
- After: hasScreen

sh1mj1 committed:
refactor(MainActivity): fragment 를 프로퍼티로 갖는다

- Before: 프래그먼트 화면 전환 시마다 인스턴스 생성

- After: 프래그먼트 인스턴스 변수를 사용할 때 초기화하고, 재사용

sh1mj1 committed:
refactor(HomeFragment): 바인딩의 메모리 누수 위험 제거

- _binding, binding 패턴 사용

sh1mj1 committed:
refactor(UIMapping.currency): 금액 형식을 strings 리소스를 사용

sh1mj1 committed:
feat(Theaters): 영화를 상영하고 있는 영화관들을 구한다

sh1mj1 committed:
refactor: 사용하지 않는 파일 제거

sh1mj1 committed:
refactor(MoneyUi): 금액이 표시되는 형식을 지정한다

sh1mj1 committed:
refactor(DateRangeUi): 날짜 범위가 표시되는 형식을 지정

sh1mj1 committed:
refactor(AdvertisementViewHolder): 광고 뷰 홀더의 파라미터 이름 변경

sh1mj1 committed:
refactor(ImageBindingAdapter): 이미지 관련 바인딩 어댑터 분리

sh1mj1 committed:
refactor(SeatsBindingAdapter): 좌석 바인딩 어댑터

sh1mj1 committed:
refactor(ScreenViewHolder): 파라미터 이름 변경

- screen -> screenPreviewUi

sh1mj1 committed:
refactor(UiMapping): ui 모델로 객체를 변경하는 로직을 이동

- UiMapping -> 각 ui 모델 클래스의 최상위

sh1mj1 committed:
refactor($ui): 사용하지 않는 파일 제거

sh1mj1 committed:
refactor(ScreenDetailPresenter): 상영 상세 프레젠터가 상영 id 와 영화관 id 를 가진다

- ScreenDetailPresenter
  - saveId 메서드 추가
  - loadScreen 메서드의 파라미터 제거
  - reserve 메서드의 파라미터 제거

sh1mj1 committed:
refactor(ScreenDetailTicketView): 중복되는 메서드 제거

sh1mj1 committed:
chore(ScreenDetailActivityTest): lint 를 위한 백틱 추가

sh1mj1 committed:
refactor(DateAdapter & TimeAdapter): 어댑터뷰의 아이템 선택 리스터를 상속한다

sh1mj1 committed:
refactor(WeeklyScreenTimePolicy): 구현체에 의존하던 것을 인터페이스에 의존하도록 함

sh1mj1 committed:
refactor(SelectDateListener -> OnItemSelectedListener): 이름 변경

- 기존 SelectDateListener 와 SelectTimeListener 모두 OnItemSelectedListener 를 쓰도록 함

sh1mj1 committed:
refactor(ScreenDetailDateTimeSpinnerView): 속성으로 가지던 어댑터 제거

sh1mj1 committed:
refactor(ScreenDetailScreenView): 이름 변경

- 인터페이스 ScreenDetailView -> ScreenDetailInformationView
- 구현체 ScreenDetailScreenView -> DefaultScreenDetailInformationView

sh1mj1 committed:
refactor(ScreenDetailActivity): 뷰의 타입을 인터페이스의 타입으로 변경

sh1mj1 committed:
refactor(HomeFragment): 생명 주기에 맞게 동작 이동

- 어댑터를 초기화하고 데이터를 로드하는 것은 onViewCreated 에서 한다

sh1mj1 committed:
refactor(HomeFragment): 리사이클러뷰 어댑터를 프래그먼트에서 연결

sh1mj1 committed:
refactor(HomeFragment): parentFragmentManager -> childFragmentManager

sh1mj1 committed:
refactor(HomePresenter): 상영과 광고를 가지는 리스트를 만드는 로직 개선

- map 사용 -> mapIndexed 사용하여 리스트를 제거할 필요 없어짐

sh1mj1 committed:
refactor(TheaterBottomSheet): 영화 목록 화면에서 영화관 시트를 보여주는 기능 개선

- HomeFragment 에서 showTheatersBottomSheet 를 호출하여 바텀 시트로 이동
- TheaterBottomSheet 에 MVP 패턴 적용
- TheaterAdapter 의 onTheaterClicked 는 theaterId 만 람다의 파라미터로 받는다
- TheaterBottomSheetPresenter 가 screenId 를 저장한다

sh1mj1 committed:
refactor(TheaterBottomSheet): 영화관 어댑터가 미리 생성된다

sh1mj1 committed:
refactor(TheaterBottomSheet): 데이터 바인딩 적용

- 프레젠터와 뷰에서 initTheaterAdapter 메서드 추가
(이전 프래그먼트에서 argument 를 받아오는 작업이 onCreate 에서 일어나기 때문에 추가)

sh1mj1 committed:
refactor(AdvertisementViewHolder): 통일성을 위한 파라미터 순서 변경

sh1mj1 committed:
refactor(ScreenAdapter): inflater 인스턴스 변수 제거

sh1mj1 committed:
refactor(ScreenDetailPresenter): 중복 사용되는 객체를 인스턴스 변수로 가짐

- dateRange, loadedScreen

sh1mj1 committed:
refactor(SeatReservationContract): 인터페이스에 맞게 메서드 이름을 추상화

sh1mj1 committed:
refactor(SeatReservationActivity): 프레젠터가 아닌 리스너를 바인딩

sh1mj1 committed:
refactor(SeatReservationPresenter): selectSeat 메서드를 여러 메서드로 분리

sh1mj1 committed:
refactor(land/activity_seat_reservation): TextView 크기 변경

sh1mj1 committed:
refactor(activity_seat_reservation): 앱 컴팻 버튼에서 버튼으로 변경

sh1mj1 committed:
refactor($layout): string resource 에 데이터 바인딩 적용

sh1mj1 committed:
refactor($presenterTest): 불필요한 mock 객체의 메서드 구현을 기본으로 설정

- relaxed = true 설정을 통해 기본으로 just runs 로 설정한다

sh1mj1 committed:
refactor(ScreenAdapter): if 문을 when 문으로 변경

sh1mj1 committed:
refactor(ScreenAdapter): 뷰 홀더를 sealed class 로 해서 어댑터 변경

sh1mj1 committed:
refactor(ScreenAd): 인터페이스가 같던 computed property 제거

- ScreenPreviewUiDiffUtil 의 로직 변경
  타입에 따라 areItemsTheSame 비교

sh1mj1 committed:
refactor(Theaters): 사용하지 않는 메서드 삭제

- screen 을 파라미터로 받는 screeningTheater 메서드 삭제

sh1mj1 commented:
라코데브, 연휴 잘 보내셨나요?!

연휴동안 가족 행사가 있어 조금 늦었습니다. 
그래도 시간을 계속 내가며 피드백 주신 부분 모두 반영했습니다.
덕분에 공부한 내용이 많아서 감사했습니다!!

### presenter 데이터 바인딩
제가 PR 을 처음 보낼 때 질문 드렸던 `presneter 를 xml 에 데이터 바인딩 해주는 것이 괜찮은가?` 에 대한 생각은 아직 변함이 없습니다.

만약 아래와 같은 xml 에 프레젠터를 매핑한다면  버튼에 `onClick= @{() -> presneter.프레젠터의 메서드}` 가 되게 됩니다. (예를 들어 `presenter.reserve()`)
![image](https://github.com/woowacourse/android-movie-theater/assets/90040304/ef2d03f7-79c5-4d6c-a210-a392c29e9bdb)

그런데 다른 화면에서 똑같은 구성의 화면을 사용하면서 버튼을 눌렀을 때의 행동이 `reserve` 가 아니라면 이 xml 을 똑같이 활용할 수 없게 됩니다.
그래서 저는 presenter 를 데이터 바인딩하는 것이 아닌, `itemClickedListener`와 같은 형태를 넘기는 것이 재사용성이 더 좋다고 생각이 들었습니다. 
이에 대해서 다시 의견 여쭙고 싶습니다!

### 모델 도메인 관련
제가 이전에 dm 보내 드린 대로 현재 도메인이 이상합니다.
현재 상영(`Screen`) 은 id, movie, dateRange(상영 시작 날짜 ~ 끝 날짜) 를 가집니다.
하지만 실제로는 아래와 같이 설계되어야 합니다.

![image](https://github.com/woowacourse/android-movie-theater/assets/90040304/788d6af6-6127-42f4-b5ba-2db291cb48c4)

왜 이렇게 설계되었는지 생각해보니, 피그마 화면을 보면서 모델 도메인을 설계하다 보니, 실제 데이터 모델 도메인이 아닌, 화면에 보여주기 위한 모델 도메인을 설계하게 된 것 같습니다.
모델 도메인을 마치 Ui 모델 도메인 처럼 설계하게 된 것 같습니다.

모델 도메인을 갈아엎지는 않았지만, 이에 대해 고민을 조금 오래 한 덕분에 다른 미션을 진행할 때는 조금 설계가 나아지지 않을까 생각이 들었습니다.

앞으로도 잘 부탁드립니다!

laco-dev commented:
> 그런데 다른 화면에서 똑같은 구성의 화면을 사용하면서 버튼을 눌렀을 때의 행동이 reserve 가 아니라면 이 xml 을 똑같이 활용할 수 없게 됩니다.
그래서 저는 presenter 를 데이터 바인딩하는 것이 아닌, itemClickedListener와 같은 형태를 넘기는 것이 재사용성이 더 좋다고 생각이 들었습니다.
이에 대해서 다시 의견 여쭙고 싶습니다!

프레젠터는 reserve()라는 기능에만 관심이 있지 클릭이던 더블클릭이던 스와이프던지는 관심이 없어요.
이미 예약한다. 라는 명확한 비즈니스 로직이 정의되어 있는데, 어떤 문제가 있을까요?

반대로 버튼을 눌렀을때의 행동이 다르다면 그건 다른겁니다. 예약을 하는 행동이 아닌 다른 행동을 하고 싶은게 아닐까요?

laco-dev commented on app/src/main/java/woowacourse/movie/domain/model/ScreenAd.kt:
```diff
@@ -1,45 +1,24 @@
-
     data class ScreenPreviewUi(
-        override val id: Int,
+        val id: Int,
```
`ScreenPreviewUi` 지금 보니 도메인에 Ui라는 키워드가 들어가있네요.
UI를 도메인 로직이라고 할 수 있나요?

UI 레이어에서 사용하기 위한 목적으로 만든 것이라면 명확하게 패키지와 파일을 나눠 주세요

sh1mj1 replied:
넵 이 부분을 제가 놓쳤군요! 감사합니다~

laco-dev commented on app/src/main/java/woowacourse/movie/domain/model/Theaters.kt:
```diff
@@ -3,10 +3,14 @@ package woowacourse.movie.domain.model
-            theaters.filter { theater ->
-                theater.screens.any { it.id == screen.id }
-            },
+            theaters.asSequence()
```
asSequence()를 사용한 목적은 무엇일까요? 단순히 빨라 보인다?

sh1mj1 replied:
처음에 filter 와 map 메서드를 사용해서 만들 때 컬렉션 연산이 2개가 중첩되어 중간 연산 결과 컬렉션을 만들지 않으려고 했었습니다. 
그래서 asSequence 를 사용했었습니다. 
sorted 와 같은 컬렉션 전체를 기반으로 처리해야 하는 연산이 있는 것도 아니어서 적합하다고 생각했습니다.

하지만 로직을 수정하면서 filter 와 any 가 되었습니다. 
지금 구조 상 여러 컬렉션 함수를 연쇄하는 게 아니라서 시퀀스를 사용하지 않아도 되겠네요.
오히려 시퀀스 사용이 더 느릴 것 같아요.
[(작은 컬렉션 처리를 한 단계만 할 때는 시퀀스가 더 느릴 수 있다 내용을 과거에 정리한 글)](https://sh1mj1-log.tistory.com/208)
링크 글 하단에 `시퀀스가 빠르지 않는 경우` 에 적어두었습니다.

laco-dev commented on app/src/main/java/woowacourse/movie/ui/DateRangeUi.kt:
```diff
@@ -0,0 +1,7 @@
+
+import woowacourse.movie.domain.model.DateRange
+
+fun dateRangeFormat(dateRange: DateRange): String = "${dateRange.start} ~ ${dateRange.endInclusive}"
```
개인적으로는 어떠한 타입으로부터 UI에 출력하기 위한 Text를 만들어 내는 목적이라면,
XXXFormatter라는 객체를 만들어서 사용하고는 합니다.

객체지향 프로그래밍에 따라 문자열로 변환하는 것도 하나의 객체가 할 역할로 볼 수 있기 때문이에요.
그러한 맥락에서 XXUtil, XXCommon 같은 것은 잘 사용하지 않는 편입니다.

sh1mj1 replied:
사실 이러한 메서드의 파일 이름을 어떻게 지어야 할지 평소에 고민이 많습니다.

이전에는 저도 XXXUiFormatter 등으로 파일 이름을 지었을 것 같습니다.
그런데 클래스를 네이밍할 때 객체가 '무엇을 하는지' 가 아니라 '무엇인지' 를 정해야 한다는 책의 내용을 보고 공감했고 미션에 적용해보려고 했습니다.

사실 책을 읽을 때도 이런 식으로 출력 Ui 를 위한 행동만 하는 클래스는 적용하기 어렵지 않나? 라는 생각을 했었는데.. 
아무래도 XXXFormatter 라는 네이밍이 최선일 것도 같군요..

laco-dev commented on app/src/main/java/woowacourse/movie/ui/ImageBindingAdapter.kt:
```diff
@@ -0,0 +1,27 @@
+    image: Image<Any>,
+) {
+    when {
+        image.imageSource is Int -> setImageResource(imageView, image.imageSource as Int)
```
Image 타입을 sealed class로 만들어 볼 수 있겠네요

sh1mj1 replied:
네! 좋은 방법이네요!!

laco-dev commented on app/src/main/java/woowacourse/movie/ui/detail/ScreenDetailContract.kt:
```diff
@@ -14,9 +13,9 @@ interface ScreenDetailContract {
-            selectDateListener: SelectDateListener,
-            selectTimeListener: SelectTimeListener,
+            screenTimePolicy: ScreenTimePolicy,
+            onDateSelectedListener: OnItemSelectedListener,
+            onTimeSelectedListener: OnItemSelectedListener,
```
MVP 패턴의 원칙에는 콜백으로 뷰와 프레젠터 사이의 커뮤니케이션을 하지 않는다도 포함되는데요.

위 기능에 대해서 대입을 해본다면 날짜 선택 -> 프레젠터로 콜백을 호출하는 것이 아닌 날짜 선택 이후 프레젠터의 인터페이스를 다시 호출하는 순서를 가지게 됩니다.

결과적으로 Presenter가 View 패키지를 알고 있어요.

sh1mj1 replied:
> 위 기능에 대해서 대입을 해본다면 날짜 선택 -> 프레젠터로 콜백을 호출하는 것이 아닌 날짜 선택 이후 프레젠터의 인터페이스를 다시 호출하는 순서를 가지게 됩니다.

이 의미가 잘 이해가 안되네요...
프레젠터로 콜백을 호출하는 것이 아닌, 뷰로 콜백을 호출해야 한다는 의미인가요?
그게 맞다면 그렇게 반영했습니다!

laco-dev commented on app/src/main/java/woowacourse/movie/ui/detail/ScreenDetailPresenter.kt:
```diff
@@ -3,33 +3,47 @@ package woowacourse.movie.ui.detail
 ) : ScreenDetailContract.Presenter {
-    private var ticket: Ticket = Ticket(MIN_TICKET_COUNT)
-    private var dateRange = DateRange(LocalDate.now(), LocalDate.now())
+    private var screenId: Int = -1
+    private var theaterId: Int = -1
```
이 값을 생성자를 통해 받아오기를 기대했는데요.
그렇게 하지 못한 이유가 있을까요?

sh1mj1 replied:
아... 이전 미션부터 Presenter 의 생성자 파라미터에 view 와 repository 만 넣다보니까, 생성자에 screenId 를 저장하면 되는 간단한 해결 방법을 생각해내지 못했네요..

laco-dev commented on app/src/main/java/woowacourse/movie/ui/home/HomeContract.kt:
```diff
@@ -2,16 +2,12 @@ package woowacourse.movie.ui.home
-            screen: Screen,
-            theaters: Theaters,
-        )
+        fun showTheatersBottomSheet(screen: Screen)
```
바텀시트를 띄울지 토스트를 띄울지 다이얼로그를 띄울지는 View가 결정할 내용 아닐까요?
만약 실제로 바텀 시트가 아닌 스낵바로 바뀌면 인터페이스도 바꿀건가요?

sh1mj1 replied:
그렇군요..
여러 Contract 의 View 인터페이스의 메서드에서 실제 구현체가 무엇인지 너무 구체적으로 나타내고 있네요..
전부 찾아서 바꿔보겠습니다.

laco-dev commented on app/src/main/java/woowacourse/movie/ui/home/HomePresenter.kt:
```diff
@@ -1,44 +1,53 @@
+        view.showScreens(screenAdList)
+    }
+
+    private fun generatedScreenAdList(
+        screenPreviewUis: MutableList<ScreenAd.ScreenPreviewUi>,
+        advertisement: ScreenAd.Advertisement,
+    ): MutableList<ScreenAd> {
```
광고를 포함한 상영 모델을 판단하고 만드는 것은 프레젠터, 모델 누구의 역할로 봐야 할까요?
흔히 프레젠터는 모델과의 협력을 통해 결과를 View에게 전파한다고 합니다.

sh1mj1 replied:
모델의 역할이겠군요!
어떠한 객체를 만들어서 책임을 부여해야 할지 네이밍을 포함해서 고민해보겠습니다

sh1mj1 replied:
`ScreensAndAdvertisementUseCase` 이라는 객체를 만들어서 Presneter 는 데이터를 받아서 보여주게끔만 처리했습니다.

하지만 현재 이 usecase 는 model 에 포함된 객체인데 ui 모델로 바꾸고 있습니다.

MoviePreviewUi 와 ScreenPreviewUi 객체의 경우 Image 를 가지고 있으며, 모델 도메인의 객체들은 그렇지 않습니다.
그리고 toPreviewUi 메서드의 경우 파라미터로 Image 를 받습니다.
```kotlin
fun Movie.toPreviewUI(image: Image<Any>) =
    MoviePreviewUI(
        title = title,
        image = image,
        runningTime = runningTime,
    )
```

```kotlin
fun Screen.toPreviewUI(image: Image<Any>) =
    ScreenAd.ScreenPreviewUi(
        id = id,
        moviePreviewUI = movie.toPreviewUI(image),
        dateRange = dateRange,
    )
```

그래서 사용하는 곳에서 MovieRepository 에게 movieId 를 통해 이미지를 요청해야 합니다.
기존에는 Presenter 에서 Ui 모델로 바꾸고 있으며 Presenter 가  MovieRepository 를 속성으로 갖고 있어서 문제가 없었으나,  이 로직이 Usecase 로 옮겨 가게 되어 모델 도메인에서 ui 모델을 만드는 모습이 되었습니다.

이를 해결하려고 모델 도메인에서 sealed interface ScreenAdDomain 를 정의해서 screen 과 advertisement 객체가 이를 상속하게 한 후, usecase 에서는  List<ScreenAdDomain> 을 리턴해주고, Presenter 에서 다시 리스트의 아이템들을 순회하면서 UiModel 로 바꿔주는 작업을 진행하다가 그만두었습니다.

model domain 에서는 관련이 없는 Screen 과 Advertisement 를 sealed interface 로 묶는 것이 적절하지 않아보였기 때문입니다.
어떻게 해결해야 할지 힌트를 주실 수 있을까요...? 😅

laco-dev commented on app/src/main/java/woowacourse/movie/ui/home/TheaterBottomSheet.kt:
```diff
@@ -4,25 +4,86 @@ import android.os.Bundle
+
+    override fun initTheaterAdapter(screen: Screen) {
+        theaterAdapter =
+            TheaterAdapter(screen) { theaterId ->
+                presenter.onTheaterSelected(theaterId)
+            }
```
TheaterAdapter가 ListAdapter로 만들어졌는데 생성자로 받는 이유가 무엇인가요?
아이템에서 계산하기 위해 필요한 것이라면 아이템에 관련된 내용이 담겨야 하지 않나요?

sh1mj1 replied:
아이고 이제야 어떻게 해결해야 할지 알겠네요. 

영화관과 상영 개수를 가지는 객체를 만들면 되겠군요.
이 쉬운 걸 왜 못했지... 감사합니다

laco-dev commented on app/src/main/java/woowacourse/movie/ui/home/TheatersBottomSheetContract.kt:
```diff
@@ -0,0 +1,30 @@
+import woowacourse.movie.domain.model.Screen
+import woowacourse.movie.domain.model.Theaters
+
+interface TheatersBottomSheetContract {
```
MVP 인터페이스와 같은 것도 결국 View가 구체적으로 어떻게 구현했는지 모르게 하기 위함인데요.
즉, 바텀시트인지 모른다는 뜻이 됩니다.

TheatersActivity가 될 수도 있고 TheatersFragment가 될 수도 있습니다.
그렇기 때문에 Contract 인터페이스에는 바텀시트 와 같은 구현체를 드러나게 작성하지 않습니다.

sh1mj1 replied:
그렇군요!
'영화를 상영하는 극장'에 대한 계약이 더 나을 것 같아요. `TheatersScreeningMovieContract`

마찬가지로 Presenter 또한 View 가 BottomSheet 일지, Activity 일지, Fragment 일지 모르기 때문에 Presenter 구현체의 이름도 바꾸는 게 좋겠어요.

laco-dev commented on app/src/main/java/woowacourse/movie/ui/home/TheatersBottomSheetContract.kt:
```diff
@@ -0,0 +1,30 @@
+    interface Presenter {
+        fun saveScreenId(screenId: Int)
+
+        fun initTheaterAdapter()
```
프레젠터가 어댑터는 왜 초기화하나요?

laco-dev commented on app/src/main/java/woowacourse/movie/ui/home/TheatersBottomSheetContract.kt:
```diff
@@ -0,0 +1,30 @@
+
+        fun loadTheaters()
+
+        fun onTheaterSelected(theaterId: Int)
```
selectTheater()랑 onTheaterSelected()는 무슨 차이가 있나요?
프레젠터라면 선택되었다 라는 콜백을 알 필요가 없지 않을까요?

sh1mj1 replied:
코드를 짜다 극장이 선택되었을 때, 어떤 동작을 해서 이렇게 지었는데 생각해보니 selectTheater 로 네이밍을 바꿔도 되겠네요

꼼꼼히 집어주셔서 감사합니다

laco-dev commented on app/src/main/java/woowacourse/movie/ui/seat/SeatReservationActivity.kt:
```diff
@@ -31,38 +29,39 @@ class SeatReservationActivity : AppCompatActivity(), SeatReservationContract.Vie
+            saveId(theaterId, timeReservationId)
+            loadAllSeats()
+            loadTimeReservation()
+            onReserveButtonClickedListener = OnReserveClickedListener { reserve(theaterId) }
+            binding.onReserveClickedListener = onReserveButtonClickedListener
```
이 코드가 Presenter의 apply 블록 내에서 호출되는게 맞나요?

sh1mj1 replied:
구조를 바꾸면서 수정했습니다!

laco-dev commented on app/src/main/java/woowacourse/movie/ui/seat/SeatReservationContract.kt:
```diff
@@ -6,15 +6,12 @@ import woowacourse.movie.domain.model.TimeReservation
 
         fun showAllSeats(seats: Seats)
 
+        fun updateTotalPrice(totalPrice: Int)
```
show()와 update()는 어떤 차이가 있나요?

sh1mj1 replied:
show 는 단순히 보여준다. update 는 기존 값에서 갱신한다 라는 의미를 가집니다.
update 라는 네이밍이 뷰에서 가지기에는 덜 추상적인 것 같습니다

laco-dev commented on app/src/main/java/woowacourse/movie/ui/seat/SeatReservationPresenter.kt:
```diff
@@ -12,76 +14,103 @@ class SeatReservationPresenter(
-            return screen
-        }.onFailure { e ->
-            throw e
+    private fun toggleSeatSelection(
+        seat: Seat,
+        seatView: View,
+    ): Boolean =
```
토글이 반환하는 값이 무엇인지 충분히 의미 전달이 되고 있나요?

sh1mj1 replied:
isChangedSeatSelection 이 더 나을 것 같네요!

laco-dev commented on app/src/main/java/woowacourse/movie/ui/seat/SeatReservationPresenter.kt:
```diff
@@ -12,76 +14,103 @@ class SeatReservationPresenter(
+        seat: Seat,
     ) {
-        view.showDialog(screenId, theaterId)
+        seatView.isSelected = true
+        selectedSeats = selectedSeats.add(seat)
     }
 
-    override fun completeReservation(
-        screenId: Int,
-        theaterId: Int,
+    private fun deselectSeat(
+        seatView: View,
+        seat: Seat,
     ) {
-        val timeReservation = reservationRepository.loadTimeReservation(screenId)
+        seatView.isSelected = false
+        selectedSeats = selectedSeats.remove(seat)
+    }
```
프레젠터에서 View를 직접 제어해도 되나요?
커스텀뷰는 UI가 아닌가요?

sh1mj1 replied:
오랫동안 고민해보았지만 완전히 해결은 아직 못했네요.. 
여기에 계속 시간을 쓸 수가 없어 여기까지만 하고 3-4 단계 구현 이후에 다시 해보겠습니다.

지금까지 변경한 결과는 아래처럼 되었는데 Presenter 가 View 의 속성을 직접 제어하지는 않지만, 메서드의 파리미터로 넘겨주는 이상한 모습이 되었어요..

```kotlin
    override fun selectSeat(
        position: Position,
        seatView: View,
    ) {
        val seat = loadedAllSeats.findSeat(position)

        if (selectedSeats.seats.size >= ticketCount) {
            view.showSelectedSeatFail(IllegalArgumentException("exceed ticket count that can be reserved."))
            return
        }
        selectedSeats = selectedSeats.add(seat)
        view.showSelectedSeat(seatView)
    }

```

laco-dev commented on app/src/main/java/woowacourse/movie/ui/seat/SeatReservationPresenter.kt:
```diff
@@ -12,76 +14,103 @@ class SeatReservationPresenter(
+    private fun loadedScreen(screenId: Int): Screen {
+        screenRepository.findById(id = screenId).onSuccess { screen ->
+            return screen
+        }.onFailure { e ->
+            throw e
+        }
```
onFailure에서 다시 쓰로우를 터트리면 Result 타입을 왜 사용한 것인가요?

sh1mj1 replied:
이 부분 수정하겠습니다!

laco-dev commented on app/src/main/res/layout/holder_screen.xml:
```diff
@@ -26,7 +30,7 @@
             app:layout_constraintStart_toStartOf="parent"
             app:layout_constraintTop_toTopOf="parent"
-            app:srcScreenAdvertisement="@{screenAd}"
+            image:imageSource="@{screenAd.moviePreviewUI.image}"
```
다시 작성하지만, 네임스페이스는 데이터바인딩을 구분하기 위한 방법이 아니에요.
image가 아니여도 아무거나 붙여도 다 동작합니다.

sh1mj1 replied:
넵. 저번 리뷰를 통해 알게 되었습니다.
그럼에도 불구하고 이렇게 한 이유는 
처음에는 화면별로 네임스페이스를 다르게 사용하면 @BindingAdapter 를 무분별하게 사용하지 않도록 권고할 수 있지 않을까 생각했습니다.

그래서 `@BindingAdapter("seats:textFormat")` 같이 만들었습니다.

그런데 image 의 경우는 화면과 같은 특정 범위에 사용하는 것이 아닌,  많은 곳에서 사용될 수 있을 것 같아 이렇게 만들었습니다.
다른 바인딩어댑터보다 더 util 느낌이라고 생각했습니다.

laco-dev reviewed (approved):
데이터바인딩을 제대로 이해하려면 MVVM 아키텍처가 들어와야 하는데요.

그렇기에 지금 MVP 구조에서는 어떻게 뷰를 재사용 할 수 있을지의 관점 아래서 MVP 아키텍처를 구현하는데 우선한다고 봐주시면 되겠습니다.

커스텀 뷰를 시도하신 방법이 나쁘지는 않지만, 커스텀 뷰를 사용하기에는 아직 기본적인 구현이 충분히 연습이 되지 않아서 피드백이 다소 분산되는 느낌이 들었습니다.

지금 작성한 커스텀 뷰에 아무런 피드백이 작성되지 않아서, 문제 없이 잘 작성되었다고 생각하실까봐 해서요.

지금 하는 미션은 안드로이드에 더 친해지기 위한 과정들이니, 도메인 위주 보다는 안드로이드 구현 중심으로 보고 있어요.
다음 미션에서도 여러 컴포넌트를 작성할 기회가 있으니 남겨드린 내용 참고해서 같이 반영 해주시면 됩니다.
